src/sentry/search/django/backend.py

```diff
@@ -8,275 +8,397 @@
+class Condition(object):
+    """\
+    Adds a single filter to a ``QuerySet`` object. Used with
+    ``QuerySetBuilder``.
+    """
+
+    def apply(self, queryset, name, parameters):
+        raise NotImplementedError
+
+
+class CallbackCondition(Condition):
+    def __init__(self, callback, destructive=True):
+        self.callback = callback
+        self.destructive = destructive
+
+    def apply(self, queryset, name, parameters):
+        accessor = parameters.pop if self.destructive else parameters.__getitem__
+        return self.callback(queryset, accessor(name))
+
+
+class ScalarCondition(Condition):
+    """\
+    Adds a scalar filter (less than or greater than are supported) to a
+    ``QuerySet`` object. Whether or not the filter is inclusive is defined by
+    the '{parameter_name}_inclusive' parameter.
+    """
+
+    def __init__(self, field, operator, destructive=True):
+        assert operator in ['lt', 'gt']
+        self.field = field
+        self.operator = operator
+        self.destructive = destructive
+
+    def apply(self, queryset, name, parameters):
+        accessor = parameters.pop if self.destructive else parameters.__getitem__
+        inclusive = accessor('{}_inclusive'.format(name))
+        return queryset.filter(**{
+            '{}__{}{}'.format(
+                self.field,
+                self.operator,
+                'e' if inclusive else ''
+            ): accessor(name)
+        })
+
+
+def get_sql_column(model, field):
+    "Convert a model class and field name to it's quoted SQL column representation."
+    return '"{}"."{}"'.format(*[
+        model._meta.db_table,
+        model._meta.get_field_by_name(field)[0].column,
+    ])
+
+
+sort_strategies = {
+    # sort_by -> Tuple[
+    #   Paginator,
+    #   String: QuerySet order_by parameter
+    # ]
+    'priority': (Paginator, '-score'),
+    'date': (DateTimePaginator, '-last_seen'),
+    'new': (DateTimePaginator, '-first_seen'),
+    'freq': (Paginator, '-times_seen'),
+}
+
+environment_sort_strategies = {
+    # sort_by -> Tuple[
+    #   String: SQL expression to generate sort value (of type T, used below),
+    #   Function[T] -> int: function for converting sort value to cursor value),
+    # ]
+    'priority': ('log(times_seen) * 600 + last_seen::abstime::int', int),
+    'date': ('last_seen', lambda score: int(to_timestamp(score) * 1000)),
+    'new': ('first_seen', lambda score: int(to_timestamp(score) * 1000)),
+    'freq': ('times_seen', int),
+}
+
+
+def get_sort_clause(sort_by):
+    engine = get_db_engine('default')
+    if engine.startswith('sqlite'):
+        return SQLITE_SORT_CLAUSES[sort_by]
+    elif engine.startswith('mysql'):
+        return MYSQL_SORT_CLAUSES[sort_by]
+    elif engine.startswith('oracle'):
+        return ORACLE_SORT_CLAUSES[sort_by]
+    elif engine in MSSQL_ENGINES:
+        return MSSQL_SORT_CLAUSES[sort_by]
+    else:
+        return SORT_CLAUSES[sort_by]
```

Comment:
`sort_strategies`, `environment_sort_strategies` and `get_sort_clause` all feel too closely coupled/dependent on each other but I haven't figured out a clean way to tie them together.